core/src/main/scala/org/apache/spark/rdd/JdbcRDD.scala

```diff
@@ -64,8 +64,8 @@ class JdbcRDD[T: ClassTag](
     // bounds are inclusive, hence the + 1 here and - 1 on end
     val length = 1 + upperBound - lowerBound
     (0 until numPartitions).map(i => {
-      val start = lowerBound + ((i * length) / numPartitions).toLong
-      val end = lowerBound + (((i + 1) * length) / numPartitions).toLong - 1
+      val start = lowerBound + ((BigDecimal(i) * length) / numPartitions).toLong
```

Comment:
Yep that's sound, good explanation. This is also why it can't overflow when `toLong` is called. I think this is the right fix with `BigInt`.